### PR TITLE
Try reporting flaky tests to issues

### PR DESCRIPTION
**Pull request by kevin940726**
### Why
As an exploration of #33809, this PR is an experiment of using GitHub issues as our flaky tests _dashboard_. It has been discussed before in #31682. The goal is to auto-retry flaky e2e tests so that they don't cause confusion to contributors and block their progress. But just retrying is not enough, the flaky failure would be hidden by the retrying system and could potentially be a bug. Instead, we report them to a GitHub issue with the detail of the failure, so that we can monitor them in the long term.

### What does it look like
Here's an example [issue](https://github.com/kevin940726/gutenberg/issues/12) I created in my fork.

There are the title of the test, the path of the test, the full error logs of the test, and the estimated flaky rate.

The issue is labeled by `flaky-test` so that we can filter them in the issues tab. Each log has a link pointing to the original failing action run.

### How does it work
The brief overview of the flow is as follow:
1. We use `jest.retryTimes(2)` to at most retry the tests 2 times (plus the original run is 3 times) before it pass.
2. We record each failure and mark the test as **flaky** if it pass after retrying using Jest's reporter API. Then, we store each flaky test result to a file and upload them to the GitHub artifact.
3. We trigger the `flaky-tests` workflow and run the `report-flaky-tests` action.
4. Inside `report-flaky-tests`, we use GitHub API to download the stored artifact, run some pre-formatting, and post the aggregated result to an issue.
5. The issue is identified by the test title, if there's already an existing issue with the same title, it will try to update the issue instead of creating a new one.

Step 4 needs the repo token to perform some of the GitHub API, due to security reason, we can't expose the token in our e2e workflow in step 2, hence we need the individual workflow and the artifacts. This is documented in a [blog post](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/) by GitHub.

The new workflow is depending on the `workflow_run` event, which will only work if the workflow file is in the default branch (`trunk`), so this PR won't work until it's merged.

The flaky rate is calculated by counting the failing times and the total amount of commits since the first recorded commit. It's only recorded in `trunk`, and it doesn't count the manual reruns, so it's just an estimated value.

## Types of changes

New feature

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/e2e-tests/config/setup-test-framework.js
Comment (tellthemachines) [inline packages/e2e-tests/config/setup-test-framework.js]: Does this just retry flaky tests, or all failing tests?

File: packages/e2e-tests/jest.config.js
Comment (noisysocks) [inline packages/e2e-tests/jest.config.js]: Should we only add this reporter when `process.env.CI` is true?

File: packages/e2e-tests/config/setup-test-framework.js
Comment (noisysocks) [inline packages/e2e-tests/config/setup-test-framework.js]: Maybe let's note here that this allows `flaky-tests-reporter` to do its thing.

File: .github/report-flaky-tests/index.js
Comment (tellthemachines) [inline .github/report-flaky-tests/index.js]: On the other hand, if original author has gone on holiday, or is otherwise absent, having the issue assigned might just stop other folks from picking it up. Maybe best leave it as is for now.

File: .gitignore
Comment (noisysocks) [inline .gitignore]: If we only run `flaky-tests-reporter` in CI then we shouldn't need to ignore this file, yeah?

File: .github/workflows/end2end-test.yml
Comment (tellthemachines) [inline .github/workflows/end2end-test.yml:54]: Wait, were we generating artifacts on _all_ test runs? 🙀

File: packages/e2e-tests/config/flaky-tests-reporter.js
Comment (tellthemachines) [inline packages/e2e-tests/config/flaky-tests-reporter.js]: Should we give a bit more context here, e.g. explain that a test that initially fails is given 2 re-runs and marked as flaky if at least one of those re-runs passes? Or something, I'm not sure I got this right 😅

File: .github/workflows/flaky-tests.yml
Comment (noisysocks) [inline .github/workflows/flaky-tests.yml]: Should the label match up with how we format other labels in the repo e.g. `[Type] Flaky Test`?

**Review by tellthemachines — COMMENTED** [review]
Thanks for working on this! I mostly left questions below 😄 

Is there any way we can test this, or shall we just merge it and test live?

File: .github/report-flaky-tests/index.js
Comment (noisysocks) [inline .github/report-flaky-tests/index.js]: Probably I don't understand how artefacts work, but won't there be multiple files because there are multiple (4) E2E jobs?

File: .github/report-flaky-tests/index.js
Comment (noisysocks) [inline .github/report-flaky-tests/index.js]: Does `listForRepo` implicitly only return open issues? If not, I can see us ending up with more than 100 issues with the flaky test label.

File: .github/report-flaky-tests/index.js
Comment (noisysocks) [inline .github/report-flaky-tests/index.js]: Could this be `return testKey.split( ':' )?`

File: .github/report-flaky-tests/index.js
Comment (noisysocks) [inline .github/report-flaky-tests/index.js:55]: I wonder if we should use a `key` that we store in the issue's metadata? Otherwise I fear somebody may accidentally rename the issue causing duplicates.

File: .github/report-flaky-tests/index.js
Comment (noisysocks) [inline .github/report-flaky-tests/index.js:62]: Should probably guard against `meta` being undefined in case somebody accidentally edits the issue.

File: .github/report-flaky-tests/index.js
Comment (noisysocks) [inline .github/report-flaky-tests/index.js:77]: I'm not understanding this. Why is the number of commits since we first reported flaky tests a good approximation for total runs? Could you give me an example?

File: packages/e2e-tests/config/setup-test-framework.js
Comment (kevin940726) [reply packages/e2e-tests/config/setup-test-framework.js]: Oh okay so this is the foundation of how we detect flaky tests in the first place. Should probably rephrase it to "retry failing tests".

- If a test fails, it will automatically re-run at most 2 times.
- If all 3 times fail, it's then reported as a **failing** test. Not a flaky test and won't report it to issue.
- If it passes after retrying, it's then marked as a **flaky** test, and will post to issue, but displayed as passed in the original commit.

File: .github/workflows/end2end-test.yml
Comment (kevin940726) [reply .github/workflows/end2end-test.yml:54]: Haha, nope. Just that we were always **uploading** the debug artifacts, whether it exists or not. If it doesn't exist (because the tests pass), but we still try to upload it, then it will display a warning. This change disable the warning.

File: .github/report-flaky-tests/index.js
Comment (noisysocks) [inline .github/report-flaky-tests/index.js:90]: If you wanted to **really** lean into the _"HTML comments are obviously the best place to store data why are you looking at me like that"_ way of life then you could use `wp.blocks.serialize()` and `wp.blocks.parse()` instead of string concatenation 😂

File: packages/e2e-tests/config/flaky-tests-reporter.js
Comment (kevin940726) [reply packages/e2e-tests/config/flaky-tests-reporter.js]: Yeah, sure! I'll provide more context like in https://github.com/WordPress/gutenberg/pull/34432#discussion_r699862384.

**Comment by kevin940726** [comment]
I think the only way we can test it for now without merging this is to create your own fork and merge this PR onto your fork's `trunk`. Might also want to create a intentionally flaky test and skip all the other tests for faster feedbacks.

File: .github/report-flaky-tests/index.js
Comment (noisysocks) [inline .github/report-flaky-tests/index.js:267]: yes

**Review by noisysocks — COMMENTED** [review]
This is freaking cool! 🎉

Should we only retry tests and report flaky tests when CI runs on `trunk`? Two reasons:

- It aligns more closely with the discussion in https://github.com/WordPress/gutenberg/pull/31682 where the consensus seems to be that we should build a report of flaky tests _before_ we automatically retry tests on PRs.
- It lets us incrementally ramp up this new bit of infrastructure. A staged rollout, if you will. This would reduce the likelihood that e.g. the job accidentally spams the repo with GitHub issues after it is deployed.

File: .github/workflows/flaky-tests.yml
Comment (kevin940726) [reply .github/workflows/flaky-tests.yml]: Yeah, most likely we should. So I made this a config rather than a hard-coded value. We'll still have to create that label first in our repo though so it's TBD 😅 .

File: .github/report-flaky-tests/index.js
Comment (kevin940726) [reply .github/report-flaky-tests/index.js]: Huh, OMG, yeah I totally forgot about that. We should aggregate them into one though.

File: .github/report-flaky-tests/index.js
Comment (kevin940726) [reply .github/report-flaky-tests/index.js]: Yep it only search for open issues, but now that I think about it we should search for all instead. So that we can re-open closed issue if it happens again. And yep, we should add pagination to fetch them all.

File: .github/report-flaky-tests/index.js
Comment (kevin940726) [reply .github/report-flaky-tests/index.js]: That won't work if the test has `:` in its title.

File: .github/report-flaky-tests/index.js
Comment (kevin940726) [reply .github/report-flaky-tests/index.js:62]: A lot of things here could break if somebody edits the issue manually 😂 , that's why I added a warning on top of the description, and I expect all members with push access should be able to read it 😅 .

File: .github/report-flaky-tests/index.js
Comment (kevin940726) [reply .github/report-flaky-tests/index.js:77]: We're only recording runs in trunk. So if we exclude manual reruns, then the total commits since the first commit that introduced the test should be the total amount of runs the test has in trunk. I can't think of a way to get the first commit that introduced the test, so the first recorded flaky commit seems like a good enough approximation for now.

File: .github/report-flaky-tests/index.js
Comment (kevin940726) [reply .github/report-flaky-tests/index.js:267]: Oh, this is actually copied from [`strip-ansi`](https://github.com/chalk/strip-ansi) but it requires ES modules, which means we need a build step. I don't want to do that for now but I'll switch to using it eventually.

File: .github/report-flaky-tests/index.js
Comment (noisysocks) [reply .github/report-flaky-tests/index.js:77]: Got it that makes sense 👍

File: .github/report-flaky-tests/index.js
Comment (kevin940726) [reply .github/report-flaky-tests/index.js:55]: Hmm, I don't understand what you mean here? How do generate such key? How do we map a test to the key?

Comment (draganescu) [reply .github/report-flaky-tests/index.js:55]: You could make a key from the title itself and then instead of saerching by title search in body for key. Key can be any hash from the title string. But I don't know about searching in body for string.

File: .github/report-flaky-tests/index.js
Comment (draganescu) [reply .github/report-flaky-tests/index.js]: I would like to stay away from auto assigning for now.

File: .github/report-flaky-tests/index.js
Comment (adamziel) [reply .github/report-flaky-tests/index.js:55]: Essentially something like:
```js
let title = "Flaky test ...";
const hash = md5(title).substr(0,6);
title = `${title} (${hash})`;
```

**Comment by adamziel** [comment]
This is such a good idea, thank you for working on that! ❤️

**Comment by talldan** [comment]
Amazing work here 👏 

> Should we only retry tests and report flaky tests when CI runs on trunk?

I did also have a concern that because PRs are a work in progress, a failing test could be in a PR, but the test could be fixed in the PR and never merged to trunk. An issue wouldn't make sense for that. That said, it would probably be a rare occurrence with the way this uses retries to detect flakiness.

On a different note, there's also an alternative model to this proposal, instead any test failure in `trunk` could be reported in an issue and wouldn't need to be retried, because by definition any failure in `trunk` is undesired.

The retries could still happen in PRs, but `trunk` is instead used as the source of truth.

File: .github/report-flaky-tests/index.js
Comment (kevin940726) [reply .github/report-flaky-tests/index.js:55]: Interesting, but like I said in https://github.com/WordPress/gutenberg/pull/34432#discussion_r699871491, I don't think it would really be a problem. Just tell everyone to not edit the title, problem solved 😆.

**Comment by kevin940726** [comment]
> On a different note, there's also an alternative model to this proposal, instead any test failure in `trunk` could be reported in an issue and wouldn't need to be retried, because by definition any failure in `trunk` is undesired.
> 
> The retries could still happen in PRs, but `trunk` is instead used as the source of truth.

This is also a cool idea, but I'm afraid if there's actually a bug in `trunk` then it will be catastrophic 😅 .

For instance, if some commit in WordPress core causes all e2e tests in gutenberg fail out of a sudden, the action will then create an issue for each of them. I think auto-retrying is a safer way to determine if a test is flaky or not. At least until we can make sure our `trunk` is stable and predictable enough 😂.

File: .github/report-flaky-tests/index.js
Comment (noisysocks) [reply .github/report-flaky-tests/index.js:55]: I thought that we had a natural key for each failing test. Re-reading this again, we don't... never mind!

File: packages/e2e-tests/config/flaky-tests-reporter.js
Comment (gziolo) [inline packages/e2e-tests/config/flaky-tests-reporter.js:7]: The interesting part in this proposal is that when the test fails but then passes with re-tries we open an issue but when the test fails all the time we do nothing. I would worry more about the latter – clearly a regression 😄

**Comment by nerrad** [comment]
Generally I really love this idea and think we should ship it and see how it works in practice on the repo. We can revert if it introduces too much noise initially. One thing that could be tried is posting comments on a single issue as part of the initial iteration to help with fine tuning frequency and content but I appreciate that might result in a lot of extra code.

File: packages/e2e-tests/config/flaky-tests-reporter.js
Comment (tellthemachines) [reply packages/e2e-tests/config/flaky-tests-reporter.js:7]: If the test fails all the time it's not considered flaky, it's just doing its job 😄 

Tests that fail consistently on CI usually get noticed and fixed pretty quickly, but tests that fail 1 out of 10 times less so.

**Comment by kevin940726** [comment]
We discussed this in our weekly core-js [meeting](https://wordpress.slack.com/archives/C5UNMSU4R/p1631027035030600). I believe we're in agreement with giving this a try.

There are still some minor issues we need to figure out before merging this though.

1. Should we run this on all PRs and trunk or only on trunk at first?
2. What should be the name of the flaky test label?
3. As proposed in https://github.com/WordPress/gutenberg/pull/34432#issuecomment-914421079, should we only post comments in one issue instead?

And here are my thoughts:

1. Let's only run this on trunk at first and iterate to all PRs when ready.
2. I'm thinking `[Type] Flaky Test`.
3. That would require a lot of changes in code but should be possible. I don't think it's worth it though. As mentioned in the original discussion in the meeting, I believe all the flaky tests have values and worth having their own corresponding issues. We can always revert this PR anytime, or several times, until we fix the noise issues (if there's any).

File: packages/e2e-tests/config/flaky-tests-reporter.js
Comment (gziolo) [reply packages/e2e-tests/config/flaky-tests-reporter.js:7]: @tellthemachines, all correct and we are on the same page here. I'm just saying that it could be also beneficial to open automatically a bur report when we detect a regression to make it easier to coordinate the flow for fixing it. In case of regression, it could also leave a message on WordPress Slack like it happens in #core channel when CI job on `trunk` fails.

Comment (kevin940726) [reply packages/e2e-tests/config/flaky-tests-reporter.js:7]: There's already the "check" itself on each commit reporting the failure of the tests. Yep, we could add some notifications or webhook if we desire, but I don't think we need to open an "issue" for them. If the tests suddenly start to fail on trunk then I think we can assume that something went wrong in our pipeline. Excluding flaky tests, the failures should have something to do with third-party libraries (core or plugins update for example). In that case, we should immediately start to fix them, opening an issue doesn't seem to help much other than creating additional noice 🤔.

And after all it's not in the scope of this PR 😅. Adding a webhook to notify failing tests on Slack is also pretty trivial with the help of the official webhook API of both GitHub and Slack.

**Comment by nerrad** [comment]
If you go the route of creating issues, one tweak that might make this nicer to follow up on is to have the bot leave a reaction on an issue that's already created. That way we can sort based on reactions (using the GitHub issue UI) to determine high flakiness tests (and prioritize which get addressed first).

I do think there needs to be some process for regularly addressing flaky tests surfaced by this new logic (either via removing, or rewriting the test) because by it's very nature a flaky test is _unreliable_ and thus can give a false impression of code coverage that doesn't actually exist. So while I still love the idea, the only concern I have is issues created by this process will pile up and not be addressed.

**Comment by nerrad** [comment]
> 1. Should we run this on all PRs and trunk or only on trunk at first?

Having this run only on trunk first seems to remove one of the main benefits of this work, which is to improve the experience for contributors where flaky tests contribute to unexpected failures in PRS where changes are unrelated to the test.

**Comment by kevin940726** [comment]
> one tweak that might make this nicer to follow up on is to have the bot leave a reaction on an issue that's already created. That way we can sort based on reactions (using the GitHub issue UI) to determine high flakiness tests (and prioritize which get addressed first).

I've thought about this, but immediately dropped it. I don't think there's an API to add multiple same reactions to an issue from the same (bot) user. I initially wanted to append new errors as comments to the issue and lock the issue, but I guess that makes the issue even less visible.

> the only concern I have is issues created by this process will pile up and not be addressed.

We already have this problem now, I don't think it'd be a new thing. Just that flaky tests are more visible now, but are also blocking contributors at the same time. The way we usually temporary fix this if we haven't come up with a proper solution for the flaky test is to simply skip it. It's the last resort and is very invisible to us (a less invisible way to overcome this is to... create an issue, which is exactly what this PR does).

> Having this run only on trunk first seems to remove one of the main benefits of this work, which is to improve the experience for contributors where flaky tests contribute to unexpected failures in PRS where changes are unrelated to the test.

Yep, agree. The end goal is always to run on all PRs. Running it on trunk is just a trial to see if it works before rolling out to all the contributors. We could also argue that it's not necessary and we can always just revert this PR though. Hence I'm open to all suggestions.

**Comment by noisysocks** [comment]
> And here are my thoughts:

I concur with all your thoughts here—let's do it! 🙂

File: .github/report-flaky-tests/index.js
Comment (noisysocks) [reply .github/report-flaky-tests/index.js:267]: Could we use one of these packages which don't seem to require ES modules?

https://www.npmjs.com/package/strip-ansi-control-characters
https://www.npmjs.com/package/strip-color

~If that's not feasible, then we need to at least add a comment above this function that contains the copyright notice.~

~https://github.com/chalk/strip-ansi/blob/main/license~

**edit:** oh, you already did that! 🙂

**Review by noisysocks — APPROVED** [review]
Let's try it! 👍

**Pull request merged**
